gas/config/tc-riscv.c

```diff
@@ -5794,3 +6096,340 @@ riscv_pop_insert (void)
 
   pop_insert (riscv_pseudo_table);
 }
+
+/* Create and get a read-only COMDAT section for an APEX instruction.
+
+   Section names encode the instruction format flags (XD=1<<0, XS=1<<1,
+   XI=1<<2, XC=1<<3), followed by the fixed Custom-0 value (11) and the
+   instruction opcode.
+
+   For example, an instruction declared as:
+     .extInstruction foo,7,XS,XC
+   maps to:
+     .riscvapex.<XS|XC>.11.<function_code>
+   and concretely:
+     .riscvapex.10.11.7
+
+   The section is placed in a COMDAT group so that the linker keeps only
+   one copy when multiple compilation units define the same section.
+   The group name is derived from the section name by removing the "riscv"
+   prefix.  */
+
+static segT
+arcv_apex_get_metadata_section (unsigned int insn_format,
+				unsigned int sub_opcode)
+{
+  const char *prefix = ".riscvapex.";
+  const char *custom_0 = "11";
+
+  /* Convert insn_format to string (max 4 chars, 0xF + null).  */
+  char insn_format_str[5];
+  sprintf (insn_format_str, "%u", insn_format);
+
+  /* Convert sub-opcode to string.  */
+  char sub_opcode_str[12];
+  sprintf (sub_opcode_str, "%u", sub_opcode);
+
+  /* Allocate buffer for full section name.  */
+  char *result = malloc (
+    strlen (prefix)
+    + strlen (insn_format_str)
+    + 1 /* dot.  */
+    + strlen (custom_0)
+    + 1 /* dot.  */
+    + strlen (sub_opcode_str));
+
+  /* Build the section name.  */
+  strcpy (result, prefix);
+  strcat (result, insn_format_str);
+  strcat (result, ".");
+  strcat (result, custom_0);
+  strcat (result, ".");
+  strcat (result, sub_opcode_str);
+
+  /* Create a new section with appropriate flags.  */
+  segT apex_section = subseg_new (result, 0);
+  bfd_set_section_flags (apex_section, SEC_READONLY | SEC_HAS_CONTENTS
+				      | SEC_LINK_ONCE);
+
+  /* Build the COMDAT group name ".apex.*" by removing "riscv" prefix.  */
+  char *group_name = malloc (strlen (result) - strlen ("riscv"));
+  /* Keep the leading dot in the group name.  */
+  group_name[0] = '.';
+  strcpy (group_name + 1, result + strlen (".riscv"));
+
+  /* Attach the group name to the section.  */
+  elf_set_group_name (apex_section, group_name);
+
+  return apex_section;
+}
+
+/* Write APEX instruction metadata into the current section.
+
+   Builds an apex_insn record with fixed fields, function code,
+   flags, and the instruction mnemonic.  The record is padded to
+   maintain 16-bit alignment, ensuring correct layout in the
+   section.  This section is later used by the disassembler to
+   decode the instruction.  */
+
+static void
+arcv_apex_write_metadata (const char *insn_name,
+			  unsigned int flags,
+			  unsigned int sub_opcode)
+{
+  struct apex_insn apex;
+  /* Compute total size including name and padding.  */
+  size_t total_size = offsetof (struct apex_insn, name)
+		      + strlen (insn_name) + 2;
+  unsigned null_padding = 1;
+
+  /* Ensure 16-bit alignment.  */
+  if (total_size & 1)
+  {
+    total_size++;
+    null_padding++;
+  }
```

Comment:
I don't really understand why `total_size` not being odd implies 16-bit aligned.